### ISSUE TITLE
[Feature Request]: Customise UI

### ISSUE THREAD

**Issue by bag-man** [puremourning/vimspector#86]
There are three things that I would love to be able to configure from my `.vimrc` or `.vimspector.json`. If these options already exist then they haven't been documented in an easy to find way, as I've spent a while looking!

1. Persist a default option for the questions `Break on all Exceptions?` and `Break on all Uncaught Exceptions?`. As right now I have to answer them everytime I launch the debugger. 
2. Configure what windows are displayed. Personally I don't use the StackTrace, Watcher and Variables windows. It would be nice if they could be disabled, to give more room for the other windows. 
3. Configure placement and size of the debug windows. It would be great to define the % splits of each window. Something similar to [Nuake](https://github.com/Lenovsky/nuake) would be awesome. 

My workflow may be a bit different to others, as I work with TypeScript, so I thought you might like to hear about how I would ideally use `vimspector` in case it is totally different to how you would. I'd like to keep it really simple, be browsing and working on code, hit a key to add break points, then another to launch the debugger. 

Rather than use watchers or variable windows, I would love to just have a responsive console (Like the JS developer console in chrome) that I could use to inspect and interact with variables. If it would be possible to have this (and the application output) open as Nuake does this would be excellent. 

This is just my personal wishlist though, I started looking at how to achieve some of these things myself, but python+vimscript is new to me so I got a little scared!

Thanks for making vimspector, it's amazing, and something that has been missing from the Vim-ecosystem for ages! :D <3

**Comment by puremourning** [comment]
For 1. You can do this with the `breakpoints` key in `.vimspector.json`. it's mentioned here: https://puremourning.github.io/vimspector/configuration.html (including an example) and came from https://github.com/puremourning/vimspector/issues/53

For the UI customisation, that would have to be roadmap. It's not something i'm likely to spend a lot of time on in the short term though. I renamed this issue to reflect that it's asking for some way to customise the UI (of which there are currently none)

**Comment by puremourning** [comment]
> I'd like to keep it really simple, be browsing and working on code, hit a key to add break points, then another to launch the debugger.

That's exactly the pattern/workflow i would expect.

> I would love to just have a responsive console (Like the JS developer console in chrome) that I could use to inspect and interact with variables.

This is what the watch window kinda gives you. In my workflow i rarely use the console and just enter expressions in the watch window, play with them, then delete them.

You enter an expression and it gives you the output in a expanding tree thingy. 

the console window is a bit lame at the moment. there's a number of reasons why, but there are also a number of existing feature requests and related discussions. e.g. https://github.com/puremourning/vimspector/issues/52

I don't imagine that Vimspector's ever going to be as good as chrome dev tools, i'm afraid.

---

aside... I don't know what Nuake is and it didn't have an obvious demo. Assuming it's just a console that pops up at the top of the window, that's what `:term`, `:vertical term`, `:botright vertical term` does in my book.

**Comment by bag-man** [comment]
Thanks for the reply! Valid points and much appreciated to get your opinion on it :) 

If I did come up with a PR to fullfill some of those features, would it be accepted?

**Comment by puremourning** [comment]
I would prefer to discuss ideas around big features like UI customisation in Gitter first, to ensure we’re aligned on goals etc and to avoid difficult or complex reviews. Also happy to review an RFC.

**Comment by bag-man** [comment]
Okay I will see what I can come up with. I think I shall start using the debugger more first to get an idea of what I might want to change better. 

One last question, is it possible to configure it to not pause on start?

Thanks again!

**Comment by puremourning** [comment]
Pause on start is a debug adapter thing. Check the config for the adapter.

**Comment by puremourning** [comment]
So here's what I'm thinking. UI code is faffy and having to code up some meta language in vim config to describe the layout it even faffier and likely to be difficult to test and difficult to maintain.

As this is a pretty advanced use case, I propose to create a `User` auto command that is triggered just after Vimspector creates it default UI. The input will be the tab page number and the window IDs of the standard window elements.

This won't be considered a stable interface (at least not for a while), so it might break you a if/when I make some changes, but I'll try and announce that on Gitter before doing it. 

Tried it out and here's an example:

```
execute 'source' expand( '<sfile>:p:h' ) . '/minimal_vimrc'
set noequalalways

function! s:CustomiseUI()
  let wins = g:vimspector_session_windows

  " Close the Variables window
  call win_execute( wins.variables, 'q' )

  " Put the stack trace at the top of the "left bar" (rotate)
  call win_gotoid( wins.stack_trace )
  wincmd r

  " Make the output window 10 lines high and right at the top of the screen
  call win_gotoid( wins.output )
  10wincmd _
  wincmd K
endfunction

augroup TestUICustomistaion
  autocmd!
  autocmd User VimspectorUICreated call s:CustomiseUI()
augroup END
```

This modifies the standard layout to some thing different:

here's a quick demo showing default vs. customised: https://asciinema.org/a/pQO87eflRTVDVW3a1TEc38HvC

There's a bug in there about the watch window being modified which I'll have to track down but it seems promising.

WDYT, world?

**Comment by BlueDrink9** [comment]
Sounds like a pretty reasonable solution for now, although I don't entirely see why this couldn't be the behind-the-scenes implementation of some options. Like, if you can manipulate the windows like this, you could use that to set the sizes, couldn't you?

**Comment by puremourning** [comment]
TL;DR setting up a UI is complex and I can't think of a way to design confutation which is sufficiently flexible and sufficiently simple and maintainable.

---

I just don't want to write a meta-UI-description language. It's basically a complete metalanguage. Consider:

* I want to make the X window larger, move it to the right of the screen and close the Y window
* I want the close the Z window, move the Y window to a new bot right vertical split and ....
* I want to have watch windows all the way down 
* If this is a python file, make the code window 80 chars and the terminal window 80 chars and close the left bar, but if this is a cpp file, make the watch window 80 chars and the left bar 80 chars. Make the stack trace 40 lines high.

Like it would require a turning complete set of operations, or would be _very_ limited. I opted for the Turing complete "run your own Vimscript code" approach as that's the most flexible and least work for me. I may even use this capability myself as I know the screen sizes and code styles that I use.

I realise that not all users are going to be that familiar with the Vim commands to do this (which is why I have some fairly sophisticated examples (you can see it's basically "jump to this window and perform some `Ctrl-W` commands"), but it's even _less_ likely that users would know some new metalanguage and I would be required to write oodles of documentation (that folk probably won't read) and would end up frustrating users.

If people would be happy with a limited set of customisation, I'd need to know what that set of operations should be. 

Happy to hear any proposals/ideas/designs considering that I'm just 1 person on this project with a full time job and another [fairly complex plugin](https://github.com/ycm-core/YouCompleteMe) to look after too.

**Comment by puremourning** [comment]
More complex example:

```viml
function! s:CustomiseUI()
  let wins = g:vimspector_session_windows

  " Close the Variables window
  if has( 'nvim' )
    " No win_execute in neovim
    call win_gotoid( wins.variables )
    quit
  else
    call win_execute( wins.variables, 'q' )
  endif

  " Put the stack trace at the top of the "left bar" (rotate)
  call win_gotoid( wins.stack_trace )
  wincmd r

  " Make the left column at least 70 chars
  70wincmd |

  " Make the code window at least 80 chars
  call win_gotoid( wins.code )
  80wincmd |

  " Make the output window 10 lines high and right at the top of the screen
  call win_gotoid( wins.output )
  10wincmd _
  wincmd K
endfunction

function s:SetUpTerminal()
  let terminal_win = g:vimspector_session_windows.terminal

  " Make the terminal window at most 80 columns wide, ensuring there is enough
  " sapce for our code window (80 columns) and the left bar (70 columns)

  " Padding is 2 for the 2 vertical split markers and 2 for the sign column in
  " the code window.
  let left_bar = 70
  let code = 80
  let padding = 4
  let cols = max( [ min( [ &columns - left_bar - code - padding, 80 ] ), 10 ] )
  call win_gotoid( terminal_win )
  execute cols . 'wincmd |'
endfunction

augroup TestUICustomistaion
  autocmd!
  autocmd User VimspectorUICreated call s:CustomiseUI()
  autocmd User VimspectorTerminalOpened call s:SetUpTerminal()
augroup END
```

**Comment by puremourning** [comment]
Turns out you can do some cools stuff with this, like customise the winbar

<img width="587" alt="Screenshot 2020-07-24 at 22 48 31" src="https://user-images.githubusercontent.com/10584846/88438107-cee9ba00-cdff-11ea-8101-4e814c888e6c.png">


I'm currently running with this:

```viml
sign define vimspectorBP text=●         texthl=WarningMsg
sign define vimspectorBPCond text=◼     texthl=WarningMsg
sign define vimspectorBPDisabled text=○ texthl=LineNr
sign define vimspectorPC text=▶         texthl=MatchParen

let g:vimspector_terminal_maxwidth = 85
let g:vimspector_bottombar_height = 25

function! s:SetUpUI() abort
  call win_gotoid( g:vimspector_session_windows.output )
  q

  call win_gotoid( g:vimspector_session_windows.code )
  nunmenu WinBar
  nnoremenu WinBar.■\ Stop :call vimspector#Stop()<CR>
  nnoremenu WinBar.▶\ Continue :call vimspector#Continue()<CR>
  nnoremenu WinBar.▷\ Pause :call vimspector#Pause()<CR>
  nnoremenu WinBar.↷\ Next :call vimspector#StepOver()<CR>
  nnoremenu WinBar.→\ Step :call vimspector#StepInto()<CR>
  nnoremenu WinBar.←\ Finish :call vimspector#StepOut()<CR>
  nnoremenu WinBar.⟲: :call vimspector#Restart()<CR>
  nnoremenu WinBar.✕ :call vimspector#Reset()<CR>
endfunction

augroup VimspectorCustom
  au!
  autocmd User VimspectorUICreated call s:SetUpUI()
augroup END
```

Looks pretty neat:

https://asciinema.org/a/R2Vogy1Lh9lswzFOOGNvK4BjH
